[숫자 야구 게임] 송여경 미션 제출합니다.

spolice0324 opened the pull request.

spolice0324 committed:
docs : 구현 기능 정리 및 흐름 정리 작성

spolice0324 committed:
feat : 게임 시작 함수 구현

spolice0324 committed:
feat : 게임 정답 문자열을 생성하는 난수 생성 함수 구현

- 게임의 정답 문자열을 만드는데 사용되는 난수 생성 함수 구현
- 제공된 Random API 사용
- 랜덤 값 추출에 Random.pickNumberInRange() 활용

spolice0324 committed:
fix : require 구문 오타 수정

- Random 을 호출하는 require 구문의 오타를 수정
- reuqire -> require

spolice0324 committed:
feat : 게임  한 턴 실행 구현

- gameTurn 메소드를 추가
- gameTurn 메소드를 통해 난수 생성을 포함하여 턴을 관리

spolice0324 committed:
docs : 구현 기능 목록 업데이트

- 1. 게임 시작과 2. 게임 한 턴 실행을 구현하였으므로 목록 업데이트

spolice0324 committed:
feat : 볼에 대한 카운트 여부를 계산하는 함수 구현

- checkBallCount
- 3. 게임 진행에 필요한 볼에 대한 카운트 여부를 계산하는 로직을 유틸 함수로 구현함

spolice0324 committed:
feat : 볼에 대한 카운트 결과 출력하기 함수 구현

- printBallCount
- 3. 게임 진행에 필요한 볼에 대한 카운트 결과 출력하는 로직을 유틸함수로 구현함

spolice0324 committed:
feat : 3. 게임 진행 함수 구현

- 입력받은 숫자에 대한 유효성 검사 로직 이외에 숫자 입력받기, 볼 카운트 계산, 출력 로직 구현
- 오답의 경우 숫자 입력 받기 재실행, 정답의 경우 정답 확인 및 종료 여부를 확인할 checkAnswer 호출

spolice0324 committed:
refactor : 모든 메소드를 비동기 형태로 변경

- async 를 사용해 비동기 메소드로 수정하여 비동기를 지원하도록 변경함
-

spolice0324 committed:
feat :  4. 정답 확인 및 종료 여부 기능 구현

- checkAnswer
- 입력 받은 숫자가 정답임을 표시, 재시작 및 완전 종료 여부 묻는 메세지 출력
- 받은 입력에 따른 조건문 추가

spolice0324 committed:
refactor : 매직 넘버 및 출력 문구 상수화

- Constants 내 매직넘버, 출력 메세지 상수로 정의함

spolice0324 committed:
refactor : 정의한 상수 값 메인 메소드 적용

- 해당 문구 및 변수를 사용하는 부분에 적용

spolice0324 committed:
refactor : 정의한 상수 값 난수 생성 함수에 적용

- 해당 문구 및 변수를 사용하는 부분에 적용

spolice0324 committed:
refactor : 정의한 상수 값 볼 카운트 출력 함수에 적용

- 해당 문구 및 변수를 사용하는 부분에 적용

spolice0324 committed:
feat : 유효성 검사 함수 구현

- 입력값에 대한 유효성 검사를 진행함

spolice0324 committed:
refactor: 볼에 대한 카운트 결과 출력 함수 로직 수정

- 비교 연산의 중복 제거
- 구조분해 할당 사용으로 가독성 향샹 및 중복 코드 정리
- 변수 명 수정

spolice0324 committed:
fix : 상수로 정의한 에러 메세지 출력 문구 변경

- 요구 조건 [ERROR] prefix 를 추가하여 ERROR_MESSAGE 문구를 수정함

spolice0324 committed:
refactor : 로직, 메소드 및 코드 구조 개선

- 정답을 생성자에서 초기화하여 클래스 변수로 관리
- 게임 재시작 및 종료 옵션을 좀 더 명확하게 처리
- 메소드 내 기능 분리로 인한 메소드 명으로 변경 checkAnswer ->  restartGame

spolice0324 committed:
feat : 유효성 검사 함수 실행 추가

- inputValidCheck 함수를 통해 입력 받은 인풋에 대한 유효성 검사를 진행

spolice0324 committed:
style : 기본 Airbnb 스타일 가이드 및 네이밍 컨벤션 준수

- makeRandomNumber -> generateRandomnumber 로 함수 명 변경
- inputValidCheck -> validateInput 으로 함수 명 변경
- 이외 스타일 가이드 적용

spolice0324 committed:
test : 난수 생성 함수 테스트 추가

- util 함수 기능 작동 테스트
- 난수를 생성하는 generateRandomNumber의 생성 조건 유효성 테스트 케이스 추가

spolice0324 committed:
test : 인풋 유효성 체크 함수 테스트 추가

- util 함수 기능 작동 테스트
- 입력값에 대해 유효성을 체크하는 테스트 케이스 추가
- 길이, 숫자, 중복, 0 조건 테스트 케이스 구분

spolice0324 committed:
test : 볼 카운트 체크 함수 테스트 추가

- util 함수 기능 작동 테스트
- 볼 카운트 체크 함수의 유효성을 체크하는 테스트 케이스 추가

spolice0324 committed:
refactor : 테스트 결과 내 undefined 출력에 따른 코드 흐름 수정

- printBallCount 함수  내 중복 제거
- 최종 메세지 출력 한번으로 수정하여 의미없는 undefined 출력을 방지함

spolice0324 committed:
test : 볼 카운트 출력 함수 테스트 추가

- util 함수 기능 작동 테스트
- 볼 카운트 출력 함수결과를 확인하는 테스트 케이스 추가

spolice0324 committed:
docs : 목록 최신화 및 구현 함수 및 메소드, 테스트 함수 정리

- 유틸함수, 메인프로그램 시작점, 유틸함수 테스트 함수에 대한 설명 추가

mi-hye commented on src/Constants.js:
```diff
@@ -0,0 +1,20 @@
+const ANSWER_LENGTH = 3;
```
아,,,몰라서 하드 코딩 했는데... 상수 파일 배우고 갑니다👍

spolice0324 replied:
코드 상에서 설명없이 등장하는 (매직 넘버) 상수나 고정적으로 사용하게 될 것들은 따로 명시해주는게 좋을 거 같습니다! 
위에서 질문하신 제스트 사용은 프리코스 준비하면서 테스트 코드 작성을 알아보면서 공부해봤습니다 !! 
이번엔 작성 안하려고 했는데 개인적으로 테스트해보다가 제가 빠뜨린 오류들을 발견해서 ... ㅎㅎ

mi-hye commented on src/utils/validateInput.js:
```diff
@@ -0,0 +1,22 @@
+import { ANSWER_LENGTH } from "../Constants";
```
예외 처리 배우고 갑니다...👍

mi-hye reviewed (commented):
커밋, readme 작성 , 코드 흐름, 변수명 등 많은 걸 배우고 갑니다.
이런 형식의 코딩이 처음인지라 구현하는데 급급해서 이런 부분들은 좀 많이 (솔직히) 대충 했었는데 이 코드 보고 많이 반성했습니다.  다른 사람에게 편하게 읽히기 위한 코드가 이런 코드군요.  2주차부터는 제대로 해야겠습니다. 감사합니다.

참고로 jest는 이번에 처음 접해봐서 잘 몰라서 그러는데 따로 공부를 하신 걸까요?

JungHun98 commented on src/utils/checkBallCount.js:
```diff
@@ -0,0 +1,20 @@
+
+  let scoreCount = { ball: 0, strike: 0 };
+
+  for (let i = 0; i < inputNumberArr.length; i++) {
+    if (inputNumberArr[i] === randomNumberArr[i]) {
+      scoreCount.strike++;
+      continue;
+    }
+    if (randomNumberArr.includes(inputNumberArr[i])) {
+      scoreCount.ball++;
+    }
+  }
```
`inputNumberArr`와 `randomNumberArr`의 길이가 같다면 배열의 고차함수를 사용해볼 수 있을 것 같아요!
여러가지 방법이 있겠지만 `reduce`를 사용한다면, 
```js
const checkBallCount = (inputNumber, randomNumber) => {

  const scoreCount = inputNumber.split("").reduce((count, digit, index) => {
    const randomDigit = randomNumber[index];

    if (digit === randomDigit) {
      count.strike++;
    } else if (randomNumber.includes(digit)) {
      count.ball++;
    }

    return count;
  }, { ball: 0, strike: 0 });

  return scoreCount;
};
```
count의 초기값을 `{ ball: 0, strike: 0 }`로 설정하고 비교 결과에 따라 값을 누적하는 방식으로도 구현해 볼 수 있을 것 같아요!!

spolice0324 replied:
for 문 대신 고차함수를 사용해야할지 고민이 있었는데 좋은 방법인 거 같네요 감사합니다 🤓

JungHun98 commented on src/Constants.js:
연관되어 있는 상수들을 따로 묶어두셨군요. 저는 분류를 하지 않았는데, 다음엔 저도 이렇게 해봐야겠어요!

haesa commented on src/App.js:
```diff
@@ -1,5 +1,58 @@
+  async gameStart() {
+    const inputValue = await Console.readLineAsync(PLAY_GAME.input);
+    if (validateInput(inputValue)) {
+      const scoreCount = checkBallCount(this.answer, inputValue);
```
`checkBallCount` 함수의 인수 순서가 뒤바뀐 것 같습니다.
함수 정의를 보니까 inputNumber가 먼저 오네요!
```js
const checkBallCount = (inputNumber, randomNumber) => {
```

spolice0324 replied:
아 그렇네요 🤔 같은 길이와 두개의 인자만 비교하니 로직상에 문제는 없지만, 다른 상황이었으면 큰 문제로 이어질 수 있겠네요 감사합니다 ㅎㅎㅎ다음엔 더 신경 써서 작성해야겠네요 !!

haesa commented on src/utils/printBallCount.js:
```diff
@@ -0,0 +1,25 @@
+  const { strike, ball } = scoreCount;
+  let message = "";
+
+  if (strike === 0 && ball === 0) {
+    message = BALL_COUNT.nothing;
+  } else {
+    if (ball !== 0) {
+      message += `${ball}${BALL_COUNT.ball}`;
+    }
+    if (strike !== 0) {
+      if (message !== "") {
+        message += " ";
+      }
+      message += `${strike}${BALL_COUNT.strike}`;
+    }
+  }
+  Console.print(message);
+  return message;
```
`if` 내부에서 바로 `message`를 출력하고 `return`으로 끝내서 각 조건을 `if`로 분리하는 것이 더 가독성에 좋을 것 같아요!
[Airbnb JS Style Guide 16.3](https://github.com/airbnb/javascript#blocks--no-else-return)

spolice0324 replied:
메세지를 출력할 부분을 분기마다 넣는게 좋았을 거 같긴하네요 🤔 링크도 참고하겠습니다 ㅎㅎ

haesa reviewed (commented):
코드 잘 보고 가요~
저도 코드리뷰 하러 와주시면 감사하겠습니다 🙏
#285

KKYHH reviewed (commented):
많이 부족한 제가 봐도 코드들이 굉장히 잘 짜여 있다고 느껴집니다 😳
보는 것만으로도 배움이 되네요 감사합니다!

혹시 테스트 코드 파일이 기존에 주어진 것과 다른데 수정해서 진행을 해도 되는 건가요?
아니면 테스트 코드 방식 중 하나일까요? 이번에 테스트 코드 개념을 처음 접해서 궁금합니다!

spolice0324 commented:
@KKYHH 
먼저 좋은 리뷰 남겨 주셔서 감사합니다 제 코드가 도움이 되었다니 ㅠㅠ... 다행이네요 ㅎㅎ ☺️

fileChanged 탭 확인하시면, 제공된 ApplicationTest 테스트 파일 이외에 UtilFunctionTest.js 파일을 보신 거 같습니다 !
제가 만든 유틸 함수들이 로직 내에서 제대로 동작하는지 확인하기 위해 작성한 코드예요! 이전에 제공된 파일은 수정하지 않았습니다 ㅎㅎ

테스트 코드 방식 중 하나인지에 대한 질문에는 어떤 답변이 적절할지 모르겠지만... 프리코스 테스트 코드는 Jest라는 JS 테스트 프레임워크를 사용하고 있어요 이 제스트 사용방법을 찾아보시면 좋을 거 같습니다!! 저는 그중 Matcher를 활용해서 테스트 대상 함수에서 반환된 값을 예측하는 케이스를 추가로 작성했어요 ㅎㅎ 도움이 되셨으면 좋겠습니다 ☺️

rlaclghks123 commented on src/utils/printBallCount.js:
```diff
@@ -0,0 +1,25 @@
+      message += `${ball}${BALL_COUNT.ball}`;
+    }
+    if (strike !== 0) {
+      if (message !== "") {
```
else내부에 if내부에 if 총 3단계의 depth가 있어서 가독성 측면도 좋지 않을 수 있고, 나중에 유지보스 측면에서도 좋지 않을 수 있지 않을까? 라는 생각이 듭니다! depth를 줄여보는 고민을 해보는건 어떨까요??

spolice0324 replied:
제가 놓쳤던 디테일인 거 같네요 다음엔 레벨을 생각하면서 중첩 구조를 만들어야겠어요 리뷰 감사합니다 ☺️

rlaclghks123 commented on src/utils/checkBallCount.js:
```diff
@@ -0,0 +1,20 @@
+      continue;
+    }
+    if (randomNumberArr.includes(inputNumberArr[i])) {
+      scoreCount.ball++;
```
단항 증감 연산자++ 보다는 +=1을 사용하면 더 좋을 것 같습니다.
[에어비엔비 컨벤션 13.6 참고](https://github.com/ParkSB/javascript-style-guide#%EB%AA%85%EB%AA%85%EA%B7%9C%EC%B9%99-naming-conventions)

spolice0324 replied:
아,, 다른 분들께서도 왜 += 연산자를 사용하셨나 싶었는데... 에어비앤비 컨벤션에 있던 내용이군요 🥹 다음 미션엔 그렇게 사용해야겠네요 !! 감사합니다

spolice0324 replied:
아,, 다른 분들께서도 왜 += 연산자를 사용하셨나 싶었는데... 에어비앤비 컨벤션에 있던 내용이군요 🥹 다음 미션엔 그렇게 사용해야겠네요 !!

rlaclghks123 commented on src/utils/validateInput.js:
```diff
@@ -0,0 +1,22 @@
+    throw new Error("[ERROR] 입력 값은 중복되는 숫자가 없어야 합니다.");
+  }
+
+  if (input.includes("0")) {
```
"0"도 상수로 분리하면 좋지 않을까 합니다!

spolice0324 replied:
정의해두는게 맞는지 고민했었는데 INVALID_INPUT 와 같이 정의해둘걸 그랬네요 조언 감사합니다 !!

rlaclghks123 reviewed (commented):
여경님 덕분에 많이 배워갑니다! 감사합니다

anttiey commented on __tests__/utilFunctionTest.js:
```diff
@@ -0,0 +1,117 @@
+    });
+  });
+
+  describe("인풋 유효성 체크 함수 테스트", () => {
```
따로 테스트 코드를 작성하셔서 유효성을 검사하셨네요! 저는 이번 과제에서는 따로 테스트 코드를 작성하지 않았는데 다음부터는 꼼꼼히 테스트 코드를 작성해 봐야겠다는 생각이 듭니다 👍

anttiey commented on src/utils/checkBallCount.js:
```diff
@@ -0,0 +1,20 @@
+  const inputNumberArr = inputNumber.split("");
+  const randomNumberArr = randomNumber.split("");
+
+  let scoreCount = { ball: 0, strike: 0 };
```
`const`로 선언한 변수가 객체 또는 배열인 경우에 할당된 내부 프로퍼티 값은 변경이 가능합니다! 객체를 선언할 때 `let`보다는 `const`를 사용한다면 더 좋을 것 같다는 생각이 들어요 🙌🏻

spolice0324 replied:
let과 const를 적절한 상황에 맞게 잘 사용하는 습관을 키워야겠네요🤔

spolice0324 replied:
앗 말씀하신대로, const로 선언된 변수가 가리키는 객체 또는 배열의 내부 프로퍼티나 요소의 값은 변경이 가능합니다만, 저는 미리 초기화 한뒤 재할당하려고 시도했기 때문에 오류가 발생하는 거 같더라구요 ! ㅜㅜ

위에서 코드리뷰 주신 것들을 더불어 로직을 조금 수정하면서, message 변수를 선언하고 값을 누적하는 방식으로 로직을 바꾸고, const로 사용할 수 있도록 개선해보겠습니다 ㅎㅎ 
let과 const를 늘 임의로 남발했던 거 같은데 다시한번 생각하는 계기가 되었네요 🤔

spolice0324 replied:
앗 말씀하신대로, const로 선언된 변수가 가리키는 객체 또는 배열의 내부 프로퍼티나 요소의 값은 변경이 가능합니다만, 저는 미리 초기화한 값으로 선언 한뒤 조건문에 따라 재할당하려고 시도했기 때문에 오류가 발생하는 거 같더라구요 ! ㅜㅜ

위에서 코드리뷰 주신 것들을 더불어 message 변수를 선언하고 값을 누적하는 방식하여 출력을 바로 하도록 분기를 수정했는데요. 여러번 재할당이 일어나 message 값이 변경되기에 let을 사용했습니다 
let과 const를 늘 임의로 남발했던 거 같은데 다시한번 생각하는 계기가 되었네요 🤔

spolice0324 replied:
앗 말씀하신대로, const로 선언된 변수가 가리키는 객체 또는 배열의 내부 프로퍼티나 요소의 값은 변경이 가능합니다. 다만, 저는 미리 초기화한 값으로 선언한 뒤 조건문에 따라 재할당하기에 오류가 발생하는 거 같더라구요 ! ㅜㅜ

위에서 코드리뷰 주신 것들을 더불어 message 변수를 선언하고 값을 누적하는 방식하여 출력을 바로 하도록 분기를 수정했는데요. 
scoreCount는 말씀하신대로 구조분해 할당으로 받아와 const로 선언하고 message에 값을 할당하는 방식으로 로직을 수정했습니다 message는 조건에 따라 재할당이 일어나 let을 사용하구요 !!
let과 const를 늘 임의로 남발했던 거 같은데 다시한번 생각하는 계기가 되었네요 🤔

anttiey reviewed (commented):
코드 잘 보고 갑니다! 논리적인 구조에 대한 고민이 느껴지는 멋진 코드인 것 같아요 😊 이미 많은 분들이 앞서 코드 리뷰를 남겨 두셔서 부족하지만 몇 개 덧붙여 봅니다 🙏

ujunglim commented on src/Constants.js:
```diff
@@ -0,0 +1,20 @@
+  input: "숫자를 입력해주세요: ",
+  check: "게임을 새로 시작하려면 1, 종료하려면 2를 입력하세요.",
+  answer: "3개의 숫자를 모두 맞히셨습니다! 게임 종료",
+};
```
하드코딩한 상수값들은 오류를 대문자를 사용하시면 가독성도 더 좋고 오류찾기에도 좋을것 같습니다~
https://www.freecodecamp.org/news/when-to-capitalize-your-javascript-constants-4fabc0a4a4c4/

ujunglim commented on src/utils/generateRandomNumber.js:
```diff
@@ -0,0 +1,16 @@
+
+  while (answer.length < ANSWER_LENGTH) {
+    const randomNumber = Random.pickNumberInRange(1, 9);
+    if (!answer.includes(randomNumber)) {
```
includes로 반복을 체크하는 것도 좋은 방법이지만 set 을 사용하셔도 좋을거 같습니다

spolice0324 replied:
set 사용도 좋은 방법이 될 수 있겠네요 조언 감사합니다 !

spolice0324 committed:
refactor : 볼 카운트 체크 함수 코드 로직 개선

- 고차함수 reduce를 사용하여 비교 결과에 따라 값을 누적하는 방식으로 변경

spolice0324 committed:
refactor : 볼 카운트 체크 함수 인자 호출 순서 변경

- inputNum, randomNum 순으로 호출

spolice0324 committed:
style : airbnb style 가이드 및 설정 prettier, eslint 적용

- single quote 사용
- import 문 수정

spolice0324 committed:
refactor : 볼 카운트 출력 부 로직 개선

- 조건 문 내부에서 메세지 출력하여 각 조건을 분리하여 가독성 증진
- Depth 2 이내로 수정
- return 하나로만 함수 로직 마무리
